test/orange/orangeinp/RevolvedPolygon.test.cc

```diff
@@ -0,0 +1,285 @@
+}
+
+//---------------------------------------------------------------------------//
+/* Test case with a single concave region.
+ \verbatim
+    3 _
+       |
+    2 _|    __________
+ z     |    \         |
+    1 _|    /         |
+       |  /           |
+    0 _|/_____________|__________
+       |    |    |    |    |    |
+       0    1    2    3    4    5
+                   r
+ \endverbatim
+ */
+TEST_F(RevolvedPolygonTest, two_levels)
+{
+    VecReal2 polygon{{1, 2}, {1.2, 1.5}, {0, 0}, {3, 0}, {3, 2}};
+
+    this->build_volume(RevolvedPolygon{"rp", std::move(polygon)});
+
+    static char const* const expected_surface_strings[]
+        = {"Plane: z=0",
+           "Plane: z=2",
+           "Cone z: t=0.5 at {0,0,0}",
+           "Cyl z: r=3",
+           "Plane: z=1.5",
+           "Cone z: t=0.8 at {0,0,0}",
+           "Cone z: t=0.4 at {0,0,4.5}"};
+    static char const* const expected_volume_strings[]
+        = {"all(+0, -1, -3, !all(+0, -1, -2), !all(!all(+0, -1, -2), "
+           "any(all(+0, -4, -5), all(-1, +4, -6))))"};
+
+    static char const* const expected_md_strings[] = {
+        "",
+        "",
+        "rp@0.0.0.mz,rp@0.0.1.mz,rp@1.0.0.mz,rp@1.0.2.mz",
+        "rp@0.0.0.pz,rp@0.0.1.pz,rp@1.0.1.pz,rp@1.0.2.pz",
+        "",
+        "rp@0.0.0.kz,rp@1.0.2.kz",
+        "",
+        "rp@0.0.0,rp@0.0.iu,rp@1.0.2,rp@1.0.iu",
+        "rp@0.0.1.cz",
+        "",
+        "rp@0.0.1,rp@0.0.ou",
+        "rp@0.0.nui,rp@1.0.nui",
+        "rp@0.0.d",
+        "rp@1.0.0.pz,rp@1.0.1.mz",
+        "",
+        "rp@1.0.0.kz",
+        "",
+        "rp@1.0.0",
+        "rp@1.0.1.kz",
+        "",
+        "rp@1.0.1",
+        "rp@1.0.ou",
+        "rp@0.cu,rp@1.0.d",
+        "rp@0.ncu",
+        "rp@0.d",
+    };
+
+    auto const& u = this->unit();
+    EXPECT_VEC_EQ(expected_surface_strings, surface_strings(u));
+    EXPECT_VEC_EQ(expected_volume_strings, volume_strings(u));
+    EXPECT_VEC_EQ(expected_md_strings, md_strings(u));
+}
+
+//---------------------------------------------------------------------------//
+/* These cases with nested concavity.
+ \verbatim
+   3 __  __ . . . . . . .  ____
+      | |  |              |    |
+   2 _| |  |     ____     |    |
+ z    | |  |    |    |    |    |
+   1 _| |  |____|. . |____|    |
+      | |______________________|
+   0 _|________________________
+      |    |    |    |    |    |
+      0    1    2    3    4    5
+                  r
+  \endverbatim
+ */
+TEST_F(RevolvedPolygonTest, three_levels)
+{
+    VecReal2 polygon{{5, 0.5},
+                     {5, 3},
+                     {4, 3},
+                     {4, 1},
+                     {3, 1},
+                     {3, 2},
+                     {2, 2},
+                     {2, 1},
+                     {1, 1},
+                     {1, 3},
+                     {0.33, 3},
+                     {0.33, 0.5}};
+
+    this->build_volume(RevolvedPolygon{"rp", std::move(polygon)});
```

Comment:
It would also be good to test a few points to make sure the constructed logic is correct (besides eyeballing it). Include `"orange/orangeinp/detail/SenseEvaluator.hh"` and do something like:
```suggestion
    auto vol_id = this->build_volume(RevolvedPolygon{"rp", std::move(polygon)});
    auto eval_sense = [&u = this->unit(), vol_id](Real3 const& point) {
        CELER_ASSERT(vol_id < u.tree.volumes().size());
        auto n = u.tree.volumes()[vol_id.get()];
        detail::SenseEvaluator eval_sense(u.tree, u.surfaces, pos);
        return eval_sense(n);
    };
```
(you could make this a class method to reuse across the different cases).